### PR TITLE
added run command

### PR DESCRIPTION
**Pull request by cheld**
Added option to execute custom shell commands (e.g.  scripts) during device init.

Syntax is similar to cloud init.

@kenan435, @zreigz

### CODE REVIEW

**Comment by Govinda-Fichtner** [comment]
@cheld This is great! Gonna have a look at it this evening!

**Comment by cheld** [comment]
@floreks could you please make a review

To run tests:
```
git clone ...
make compile
make test
make test-shell
```
@Govinda-Fichtner do you need anything to have the PR merged?

File: README.md [hunk 1]
```diff
@@ -47,6 +47,14 @@ docker:
     - "/path/to/another-image-name.tar"
 ```
 
+### Run a Command
+device-init can execute a list of commands on boot. e.g.:
+```yaml
+runcmd:
+  - "apt-get install package-name"
```
Comment (floreks) [inline README.md]: According to tests it actually works without quotes in yaml, so we can skip them in here. Go run command should escape this string.

File: cmd/runcommand.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
+package cmd
```
Comment (floreks) [inline cmd/runcommand.go]: I'm not sure about this project formatting rules but you could use gofmt on this file.

File: README.md (hunk 1 above)
Comment (cheld) [inline README.md]: hmmm...cloud init documents examples without quotes. That's why I added tests for it. On the other hand the internet is full of forum post of users asking for help, because they did not realize that their unquoted commands produces incorrect yaml.
So, to keep things save and simple I would suggest not to document that quotes can be removed in some cases

Comment (floreks) [inline README.md]: In this case yaml in unmarshalled in go code and shell command is run also from go code. These issues can happen when running in console, i.e.:

```bash
# Standard command execution
14:52 $ ls -la | grep test
-rw-rw-r-- 1 floreks floreks  895 lut 22 14:52 test.go

# Using 'sh -c' without quotng command argument
14:52 $ sh -c ls -la | grep test
test.go

# Using 'sh -c' with quoting command argument
14:53 $ sh -c "ls -la | grep test"
-rw-rw-r-- 1 floreks floreks  895 lut 22 14:52 test.go
```

However when we use go code to execute these commands we do not rely on standard console behavior. Check out this [go example](https://play.golang.org/p/KY_4S9CF8_) (it won't run in playground because of additional import).

Output:
```
Quoted command: ls -la | grep test
Unquoted command: ls -la | grep test

--------------------------------
Quoted command out:
-rw-rw-r-- 1 floreks floreks  895 lut 22 14:52 test.go

--------------------------------
Unquoted command out:
-rw-rw-r-- 1 floreks floreks  895 lut 22 14:52 test.go
```

As you can see in both cases array object is converted into a single string object and quotes are dropped. 

`config.GetStringSlice` method is part of viper package that uses same yaml package for unmarshalling. In this case there is no risk related to unquoted commands.

Comment (cheld) [inline README.md]: interesting - my assumption was that `ls -la | grep test` will not work because of the pipe symbol. I tried some more examples and it is quite hard to produce yaml syntax errors. But after trying hard enough it is possible e.g.
```
runcmd:
 - echo hello: > /tmp/test-without-quotes.txt
```
this yaml has syntax error because of colon followed by space (you can also check on http://www.yamllint.com/), while 
```
runcmd:
 - "echo hello: > /tmp/test-with-quotes.txt"
```
works.

so, at least I will change the test to make sure it requires the quotes and try to make the tests more descriptive (e.g. adding comments).

File: cmd/runcommand.go (hunk 2 above)
Comment (Govinda-Fichtner) [inline cmd/runcommand.go]: @cheld @floreks I finally found time to have a look at it. Sorry guys, for not being able to look at it earlier.

Compiling and running the tests works for me.

When I run the examples of the test manually via `make test-shell` I get the following output:

```
root@device-tester:/tmp# device-init -c
Using config file: /boot/device-init.yaml
Could not parse string to boolean value strconv.ParseBool: parsing "": invalid syntax
root@device-tester:/tmp# echo $?
0
root@device-tester:/tmp# ls
test-with-quotes.txt  test-without-quotes.txt
```

As you can see the exit code is fine and the two expected files are present.
Still it would be nice to get rid of the "Could not parse string to boolean value strconv.ParseBool: parsing "": invalid syntax" part. I do not know if this was introduced by this PR or not. Would have to check.

Comment (cheld) [inline cmd/runcommand.go]: The error is caused by cluster_lab.go (https://github.com/hypriot/device-init/blob/master/cmd/cluster_lab.go#L54)

A check for nil or empty string should be added, I guess. Should I fix?

File: README.md (hunk 1 above)
Comment (floreks) [inline README.md]: Interesting. My assumption was that yaml file should have correct syntax, otherwise we might eventually end up with error while decoding the file. I guess it depends on the yaml library used underneath and how it handles such cases.

Comment (cheld-demo) [inline README.md]: The device-init.yaml is provided by end-user. So, he is responsible for providing a correct yaml that can be parsed

Comment (cheld) [inline README.md]: The user is providing the device-init.yaml. So, he is responsible that the yaml is correct and can be parsed. 

The important part is that it is easy for the end user to write command lines without too much yaml-escaping.

File: cmd/runcommand.go (hunk 2 above)
Comment (cheld) [inline cmd/runcommand.go]: I think the error should be fixed in separate PR

**Comment by cheld** [comment]
@Govinda-Fichtner PTAL

File: cmd/runcommand.go (hunk 2 above)
Comment (Govinda-Fichtner) [inline cmd/runcommand.go]: @cheld I agree. Thanks for taking a look.

**Comment by Govinda-Fichtner** [comment]
@cheld Thanks you for taking the time to create this PR!
